server/src/main/java/org/apache/druid/server/coordinator/ServerHolder.java

```diff
@@ -337,29 +347,42 @@ private SegmentAction simplify(SegmentAction action)
     return action == SegmentAction.REPLICATE ? SegmentAction.LOAD : action;
   }
 
-  private boolean updateQueuedSegments(DataSegment segment, SegmentAction action, boolean addToQueue)
+  private void addToQueuedSegments(DataSegment segment, SegmentAction action)
   {
-    if (addToQueue) {
-      queuedSegments.put(segment, action);
+    queuedSegments.put(segment, action);
+
+    // Add to projected if load is started, remove from projected if drop has started
+    if (action.isLoad()) {
+      projectedSegments.add(segment);
     } else {
-      queuedSegments.remove(segment);
+      projectedSegments.remove(segment);
     }
 
-    final long sizeDelta = addToQueue ? segment.getSize() : -segment.getSize();
     if (action.isLoad()) {
-      sizeOfLoadingSegments += sizeDelta;
+      sizeOfLoadingSegments += segment.getSize();
     } else if (action == SegmentAction.DROP) {
-      sizeOfDroppingSegments += sizeDelta;
+      sizeOfDroppingSegments += segment.getSize();
+    } else {
+      // MOVE_FROM actions graduate to DROP after the corresponding MOVE_TO has finished
+      // Do not consider size delta until then, otherwise we might over-assign the server
     }
+  }
```

Comment:
After the change to remove teh `addToQueue` boolean, you now have the exact same if clause at the beginning of 2 subsequence if statements.  You should be able to make it

```
if (action.isLoad()) {
  projectedSegments.add(segment);
  sizeOfLoadingSegments += segment.getSize();
} else {
  projectedSegments.remove(segment);
  // The current else block could be from a DROP or a MOVE_FROM.  The MOVE_FROM will eventually graduate to
  // DROP after ...
  if (action == SegmentAction.DROP) {
    sizeOfDroppingSegments += segment.getSize()
  }
}
```